dev-notes/minimal-os-for-fobnail.md

```diff
@@ -0,0 +1,718 @@
+Minimal OS would communicate with Fobnail Token, perform attestation in order
+to determine platform state, and then hand the control back to the main OS.
+Currently, only early launch will be supported, late launch is planned as a
+future improvement.
+
+## OS requirements
+
+We pay attention to the drivers supported by each OS. We need the following
+drivers:
+
+- TPM drivers
+- Drivers required for communicating with Fobnail Token
+  - USB host drivers - at minimum EHCI and xHCI support, however UHCI and OHCI
+    would be good to have
+  - USB EEM driver - emulated Ethernet driver
+  - Network stack with IPv4 support
+
+Also we pay attention to OS security, ability to run in DLME, including
+portability between different hardware (a single binary should be able to boot
+on all x86 platforms with support for ACPI) and ability to chainload another,
+target OS.
+
+The table below summarises OS features which are a hard requirement, though they
+could be implemented by us.
+
+| Requirement             | Description                                                                    |
+| ----------------------- | ------------------------------------------------------------------------------ |
+| USB host driver         | Required for communication with Fobnail Token                                  |
+| USB EEM driver          | Required for communication with Fobnail Token                                  |
+| Network Stack           | Required for communication with Fobnail Token                                  |
+| TPM driver              | Required to perform attestation                                                |
+| Bootloader Capabilities | Required to boot target OS                                                     |
+| C library               | [Fobnail Attester](https://github.com/fobnail/fobnail-attester) is writen in C |
+| Bootable by SKL         | Whether OS can be loaded by TrenchBoot SKL without SKL or OS modification      |
+
+These are another features which are taken into account (soft requirements).
+
+| Feature                  | Description                                                               |
+| ------------------------ | ------------------------------------------------------------------------- |
+| Microkernel              | Microkernels are more secure and are preferred                            |
+| OS portability           | Required to avoid rebuilding minimal OS for each device                   |
+| CPU Architecture support | OS supported architectures, mostly we consider x86, ARM, RISC-V and POWER |
+
+### OS score counting
+
+The table at the bottom of this document contains summary of features of all
+OSes together with theirs scores. Each score is computed using the following
+rules
+
+- If feature is present then +1
+- If feature is not present then 0, or if feature would be hard to implement
+  then -1
+- If this is a hard requirement then multiple result by 2
```

Comment:
Scoring explanation should be neat the table.

Reply:
I move the whole scoring section to summary.